includes/Modules/Analytics/Advanced_Tracking/Measurement_Events/Measurement_Event.php

```diff
@@ -0,0 +1,146 @@
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Modules\Analytics\Advanced_Tracking\Measurement_Events;
+
+use \Exception;
+
+/**
+ * Class for representing a single tracking event that Advanced_Tracking tracks.
+ *
+ * @since n.e.x.t.
+ * @access private
+ * @ignore
+ */
+final class Measurement_Event implements \JsonSerializable {
+
+	/**
+	 * The measurement event's configuration.
+	 *
+	 * @since n.e.x.t.
+	 * @var array
+	 */
+	private $config;
+
+	/**
+	 * Measurement_Event constructor.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param array $config The event's configuration.
+	 * @throws \Exception Thrown when config param is undefined.
+	 */
+	public function __construct( $config ) {
+		$this->validate_config_keys( $config );
+		$this->validate_config_value_types( $config );
+
+		$this->config = $config;
+	}
+
+	/**
+	 * Validates configuration keys.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param array $config The event's configuration.
+	 * @throws \Exception Thrown when duplicate keys or invalid keys.
+	 */
+	private function validate_config_keys( $config ) {
+		$valid_keys = array(
+			'pluginName' => false,
+			'category'   => false,
+			'action'     => false,
+			'selector'   => false,
+			'on'         => false,
+			'metadata'   => false,
+		);
+		foreach ( $config as $key => $value ) {
+			if ( array_key_exists( $key, $valid_keys ) ) {
+				if ( $valid_keys[ $key ] ) {
+					throw new \Exception( 'Duplicate configuration parameter: ' . $key );
+				} else {
+					$valid_keys[ $key ] = true;
+				}
+			} else {
+				throw new \Exception( 'Invalid configuration parameter: ' . $key );
+			}
+		}
+	}
+
+	/**
+	 * Validates configuration value types.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @param array $config The event's configuration.
+	 * @throws \Exception Thrown when configuration value type is invalid.
+	 */
+	private function validate_config_value_types( $config ) {
+		foreach ( $config as $key => $value ) {
+			$value_type = gettype( $value );
+			if ( 'string' !== $value_type ) {
+				throw new \Exception( 'Invalid type [' . $value_type . '] for configuration paramter: ' . $key );
+			}
+		}
+	}
+
+	/**
+	 * Returns event configuration selector as AMP trigger name.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @return string The event selector.
+	 */
+	public function get_amp_trigger_name() {
+		return $this->config['selector'];
+	}
+
+	/**
+	 * Converts the event configuration to an AMP configuration.
+	 *
+	 * @since n.e.x.t.
+	 *
+	 * @return array $amp_config The AMP configuration for this event.
+	 */
+	public function to_amp_config() {
+		$amp_config             = array();
+		$amp_config['selector'] = $this->config['selector'];
+		$amp_config['on']       = $this->config['on'];
+
+		$vars_config                   = array();
+		$vars_config['event_name']     = $this->config['action'];
+		$vars_config['event_category'] = $this->config['category'];
+
+		$amp_config['vars'] = $vars_config;
+		return $amp_config;
```

Comment:
Clearly we don't get the metadata here for AMP. That's something I was already expecting to not be possible since we can't run custom JS in AMP, but I wonder whether they offer an alternative mechanism which maybe at least gives us limited abilities to determine things. I wonder if you can pass another selector somehow to get a value, or some kind of template variable that will be automatically replaced or so. May be worth looking into, although it's certainly possible we _may_ not find anything suitable.

Reply:
We'll look into it!

Reply:
With the PHP approach we've been talking about this should be possible.